File: cmd/api/app/types/api/v1/cluster.go
/*
Copyright 2024 The Karmada Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package v1

import (
	"github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
	corev1 "k8s.io/api/core/v1"
)

// PostClusterRequest is the request body for creating a cluster.
// PostClusterRequest 是创建集群的请求
type PostClusterRequest struct {
	// MemberClusterKubeConfig 是成员集群的 kubeconfig
	MemberClusterKubeConfig string                   `json:"memberClusterKubeconfig" binding:"required"`
	// SyncMode 是集群同步模式
	SyncMode                v1alpha1.ClusterSyncMode `json:"syncMode" binding:"required"`
	// MemberClusterName 是成员集群的名称
	MemberClusterName       string                   `json:"memberClusterName" binding:"required"`
	// MemberClusterEndpoint 是成员集群的端点
	MemberClusterEndpoint   string                   `json:"memberClusterEndpoint"`
	// MemberClusterNamespace 是成员集群的命名空间
	MemberClusterNamespace  string                   `json:"memberClusterNamespace"`
	// ClusterProvider 是集群提供商
	ClusterProvider         string                   `json:"clusterProvider"`
	// ClusterRegion 是集群区域
	ClusterRegion           string                   `json:"clusterRegion"`
	// ClusterZones 是集群区域
	ClusterZones            []string                 `json:"clusterZones"`
}

// PostClusterResponse is the response body for creating a cluster.
// PostClusterResponse 是创建集群的响应
type PostClusterResponse struct {
}

// LabelRequest is the request body for labeling a cluster.
// LabelRequest 是标签集群的请求
type LabelRequest struct {
	// Key 是标签的键
	Key   string `json:"key"`
	// Value 是标签的值
	Value string `json:"value"`
}

// TaintRequest is the request body for tainting a cluster.
// TaintRequest 是污点集群的请求
type TaintRequest struct {
	// Effect 是污点的效果
	Effect corev1.TaintEffect `json:"effect"`
	// Key 是污点的键
	Key    string             `json:"key"`
	// Value 是污点的值
	Value  string             `json:"value"`
}

// PutClusterRequest is the request body for updating a cluster.
// PutClusterRequest 是更新集群的请求
type PutClusterRequest struct {
	// Labels 是标签
	Labels *[]LabelRequest `json:"labels"`
	// Taints 是污点
	Taints *[]TaintRequest `json:"taints"`
}

// PutClusterResponse is the response body for updating a cluster.
// PutClusterResponse 是更新集群的响应
type PutClusterResponse struct{}

// DeleteClusterRequest is the request body for deleting a cluster.
// DeleteClusterRequest 是删除集群的请求
type DeleteClusterRequest struct {
	// MemberClusterName 是成员集群的名称
	MemberClusterName string `uri:"name" binding:"required"`
}

// DeleteClusterResponse is the response body for deleting a cluster.
// DeleteClusterResponse 是删除集群的响应
type DeleteClusterResponse struct {
}

File: ui/apps/dashboard/src/services/cluster.ts
/*
Copyright 2024 The Karmada Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { IResponse, karmadaClient } from './base';

export interface ObjectMeta {
  name: string;
  creationTimestamp: string;
  uid: string;
  labels?: Record<string, string>;
}

export interface TypeMeta {
  kind: string;
}

export interface NodeSummary {
  totalNum: number;
  readyNum: number;
}

export interface AllocatedResources {
  cpuCapacity: number;
  cpuFraction: number;
  memoryCapacity: number;
  memoryFraction: number;
  allocatedPods: number;
  podCapacity: number;
  podFraction: number;
}

export interface Cluster {
  objectMeta: ObjectMeta;
  typeMeta: TypeMeta;
  ready: string;
  kubernetesVersion: string;
  syncMode: 'Pull' | 'Push';
  nodeSummary: NodeSummary;
  allocatedResources: AllocatedResources;
}

export interface ListClusterResp {
  listMeta: {
    totalItems: number;
  };
  clusters: Cluster[];
  errors: string[];
}

export interface ClusterDetail extends Cluster {
  taints: TaintParam[];
}

// 新增：节点相关接口
export interface ClusterNode {
  objectMeta: ObjectMeta;
  typeMeta: TypeMeta;
  ready: boolean;
  allocatedResources?: {
    cpuCapacity: number;
    cpuFraction: number;
    memoryCapacity: number;
    memoryFraction: number;
    podCapacity: number;
    podFraction: number;
    allocatedPods: number;
  };
  resourceSummary?: {
    cpu?: {
      capacity: string;
      allocatable: string;
      allocated: string;
      utilization: string;
    };
    memory?: {
      capacity: string;
      allocatable: string;
      allocated: string;
      utilization: string;
    };
    pods?: {
      capacity: string;
      allocatable: string;
      allocated: string;
      utilization: string;
    };
  };
  status?: {
    capacity?: {
      cpu: string;
      memory: string;
      pods: string;
    };
    allocatable?: {
      cpu: string;
      memory: string;
      pods: string;
    };
    conditions?: Array<{
      type: string;
      status: string;
      lastTransitionTime: string;
      reason: string;
      message: string;
    }>;
    addresses?: Array<{
      type: string;
      address: string;
    }>;
    nodeInfo?: {
      machineID: string;
      systemUUID: string;
      bootID: string;
      kernelVersion: string;
      osImage: string;
      containerRuntimeVersion: string;
      kubeletVersion: string;
      kubeProxyVersion: string;
      operatingSystem: string;
      architecture: string;
    };
  };
}

export interface ListMemberClusterNodesResp {
  listMeta: {
    totalItems: number;
  };
  nodes: ClusterNode[];
}

export async function GetClusters() {
  const resp = await karmadaClient.get<IResponse<ListClusterResp>>('/cluster');
  return resp.data;
}

export async function GetClusterDetail(clusterName: string) {
  const resp = await karmadaClient.get<IResponse<ClusterDetail>>(
    `/cluster/${clusterName}`,
  );
  return resp.data;
}

export async function CreateCluster(params: {
  kubeconfig: string;
  clusterName: string;
  mode: 'Push' | 'Pull';
}) {
  // /api/v1/cluster
  const resp = await karmadaClient.post<IResponse<string>>(`/cluster`, {
    memberClusterKubeconfig: params.kubeconfig,
    memberClusterName: params.clusterName,
    syncMode: params.mode,
  });
  return resp.data;
}

export interface LabelParam {
  key: string;
  value: string;
}

export interface TaintParam {
  key: string;
  value: string;
  effect: string;
}

export async function UpdateCluster(params: {
  clusterName: string;
  labels: LabelParam[];
  taints: TaintParam[];
}) {
  const { clusterName, ...restParams } = params;
  const resp = await karmadaClient.put<IResponse<ClusterDetail>>(
    `/cluster/${clusterName}`,
    restParams,
  );
  return resp.data;
}

export async function DeleteCluster(clusterName: string) {
  const resp = await karmadaClient.delete<IResponse<ClusterDetail>>(
    `/cluster/${clusterName}`,
  );
  return resp.data;
}

// 新增：获取成员集群节点列表
export async function GetMemberClusterNodes(params: { clusterName: string }) {
  const resp = await karmadaClient.get<IResponse<ListMemberClusterNodesResp>>(
    `/member/${params.clusterName}/nodes`,
  );
  return resp.data;
}
